go/runtime/bundle/registry.go

```diff
@@ -0,0 +1,346 @@
+
+// Registry is an interface for handling newly discovered bundles.
+type Registry interface {
+	// HasBundle returns true iff the registry has the bundle.
+	HasBundle(manifestHash hash.Hash) bool
+
+	// AddBundle adds a bundle from the given path.
+	AddBundle(path string, manifestHash hash.Hash) error
+
+	// GetVersions returns versions for the given runtime.
+	GetVersions(runtimeID common.Namespace) []version.Version
+
+	// WatchVersions provides a channel that streams runtime versions as they
+	// are added to the registry.
+	WatchVersions(runtimeID common.Namespace) (<-chan version.Version, *pubsub.Subscription)
+
+	// GetManifests returns all known manifests that contain RONL component.
+	GetManifests() []*Manifest
+
+	// GetName returns optional human readable runtime name.
+	GetName(runtimeID common.Namespace, version version.Version) (string, error)
+
+	// GetComponents returns components for the given runtime and version.
+	GetComponents(runtimeID common.Namespace, version version.Version) ([]*ExplodedComponent, error)
+}
+
+// registry is a registry of runtime bundle manifests and components.
+type registry struct {
+	mu sync.RWMutex
+
+	dataDir string
+
+	bundles    map[hash.Hash]struct{}
+	manifests  map[common.Namespace]map[version.Version]*Manifest
+	components map[common.Namespace]map[component.ID]map[version.Version]*ExplodedComponent
+	notifiers  map[common.Namespace]*pubsub.Broker
+
+	logger *logging.Logger
+}
+
+// NewRegistry creates a new bundle registry, using the given data directory
+// to store the extracted bundle files.
+func NewRegistry(dataDir string) Registry {
+	logger := logging.GetLogger("runtime/bundle/registry")
+
+	return &registry{
+		dataDir:    dataDir,
+		bundles:    make(map[hash.Hash]struct{}),
+		manifests:  make(map[common.Namespace]map[version.Version]*Manifest),
+		components: make(map[common.Namespace]map[component.ID]map[version.Version]*ExplodedComponent),
+		notifiers:  make(map[common.Namespace]*pubsub.Broker),
+		logger:     logger,
+	}
+}
+
+// HasBundle implements Registry.
+func (r *registry) HasBundle(manifestHash hash.Hash) bool {
+	r.mu.RLock()
+	defer r.mu.RUnlock()
+
+	_, ok := r.bundles[manifestHash]
+	return ok
+}
+
+// AddBundle implements Registry.
+func (r *registry) AddBundle(path string, manifestHash hash.Hash) error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	r.logger.Info("adding bundle",
+		"path", path,
+		"manifest_hash", manifestHash,
+	)
+
+	// Open the bundle and release resources when done.
+	bnd, err := Open(path)
+	if err != nil {
+		return fmt.Errorf("failed to open bundle '%s': %w", path, err)
+	}
+	defer bnd.Close()
+
+	// Verify manifest hash.
+	if !bnd.manifestHash.Equal(&manifestHash) {
+		return fmt.Errorf("invalid manifest hash (got: '%s', expected: '%s')",
+			bnd.manifestHash.Hex(),
+			manifestHash.Hex(),
+		)
+	}
+
+	// Skip already processed bundles.
+	if _, ok := r.bundles[manifestHash]; ok {
+		return nil
+	}
+
+	// Verify that components are unique.
```

Comment:
```suggestion
	// Ensure the manifest doesn't include a component version already in the registry.
```

Something like that sounds clearer to me. Otherwise, I thought we were checking that the manifest doesn't contain duplicate components (which is already checked in `Open`).

Reply:
Agree with the suggested comment.

Question: Why is it problematic to receive a new bundle (defined by manifest), that has a component whose version is already in the registry? Isn't this too prohibitive as it does not allow to (re)-use existing components?

Reply:
What do you do when two components have the same version but different values (e.g. `Disabled`)? 
For now, we can probably be more strict, and we can loosen this latter.

Reply:
Good point. Feel free to close from my side.